grin/src/Pipeline/Definitions.hs

```diff
@@ -0,0 +1,180 @@
+{-# LANGUAGE TemplateHaskell, PatternSynonyms #-}
+module Pipeline.Definitions where
+
+import System.FilePath
+
+import Data.Text (Text)
+
+import Control.Monad.Trans.Reader
+import Control.Monad.Trans.State.Strict
+
+import Lens.Micro.Platform
+import Lens.Micro.Internal
+
+import Text.PrettyPrint.ANSI.Leijen (Doc)
+
+import Grin.Grin
+import Grin.EffectMap
+import Grin.TypeEnvDefs
+import AbstractInterpretation.Sharing
+import AbstractInterpretation.HeapPointsTo
+import AbstractInterpretation.CreatedBy    
+import AbstractInterpretation.LiveVariable
+import AbstractInterpretation.HPTResult
+import AbstractInterpretation.CByResultTypes    
+import AbstractInterpretation.LVAResultTypes
+
+data Transformation
+  -- Simplifying
+  = RegisterIntroduction
+  | ProducerNameIntroduction
+  | Vectorisation
+  | SplitFetch
+  | CaseSimplification
+  | RightHoistFetch
+  | InlineEval
+  | InlineApply
+  | InlineBuiltins
+  -- Misc
+  | GenerateEval
+  | BindNormalisation
+  | ConstantFolding
+  | UnitPropagation
+  | MangleNames
+  -- Optimizations
+  | EvaluatedCaseElimination
+  | TrivialCaseElimination
+  | SparseCaseOptimisation
+  | UpdateElimination
+  | NonSharedElimination
+  | CopyPropagation
+  | ConstantPropagation
+  | DeadCodeElimination
+  | DeadDataElimination
+  | DeadFunctionElimination
+  | DeadParameterElimination
+  | DeadVariableElimination
+  | SimpleDeadFunctionElimination
+  | SimpleDeadVariableElimination
+  | SimpleDeadParameterElimination
+  | CommonSubExpressionElimination
+  | CaseCopyPropagation
+  | CaseHoisting
+  | GeneralizedUnboxing
+  | ArityRaising
+  | LateInlining
+  deriving (Enum, Eq, Ord, Show)
+
+newtype Hidden a = H a
+
+instance Show (Hidden a) where
+  show _ = "(hidden)"
+
+instance Eq (Hidden a) where
+  _ == _ = True
+
+data AbstractComputationStep
+  = CompileToAbstractProgram
+  | OptimiseAbstractProgram
+  | PrintAbstractProgram
+  | RunAbstractProgramPure
+  | PrintAbstractResult
+  deriving (Eq, Show)
+
+data SharingAnalysisStep
```

Comment:
Please remove as sharing should be part of the HPT run.

Reply:
Now the sharing analysis will be a separate data-flow analysis. Similarly to created-by, it extends the heap-points-to analysis.